File: Matlab/LBD/findPathCoM.m
function [pathCoM, pathToAtlasIndex] = findPathCoM(pathLUT,AtlasToPathLUT,AtlasCoM)



%NS=size(thicknessVals,1);
NL=height(AtlasToPathLUT);

%matchedThicknessData=nan(NS,NL,2);
%matchedPathData=nan(NS,NL,2);
pathCoM=zeros(NL,3,2);
pathToAtlasIndex = cell(NL,2); 
hemi = {'L','R'}

        for l = 1:NL
            for r=1:2
            AOName = AtlasToPathLUT.Matched{l};
            
            pathLabelIdx = find(matches(pathLUT.PathName,AOName) & matches(pathLUT.Label_Hemi,hemi{r}));
            if(~isempty(pathLabelIdx))
                atlasIdx = pathLUT.Label_ID400x7(pathLabelIdx);
             
                pathCoM(l,:,r)=mean(AtlasCoM(atlasIdx,:),1,'omitnan');
                pathToAtlasIndex{l,r}=atlasIdx;
                    
                %giiSurf=GII.GII.giiSurface_Both;
                %valDisp=100*(ismember(GII.GII.cdata,atlasIdx));
                %figure(1)
                %clf
                %t=trisurf(giiSurf.faces,giiSurf.vertices(:,1),giiSurf.vertices(:,2),giiSurf.vertices(:,3),valDisp(:),'EdgeColor','none');
                %alpha(t,.05)
 %               for s = 1:NS
                    %matchedThicknessData(s,l,r) = mean(thicknessVals(s,atlasIdx),2,'omitnan');
                    %matchedPathData(s,l,r) = mean(pathVals(s,atlasIdx),2,'omitnan');
                    
                    %hold on;
                    %scatter3(meanloc(1))
                    %scatter3(AtlasCoM(atlasIdx,1),AtlasCoM(atlasIdx,2),AtlasCoM(atlasIdx,3))
                    %scatter3(pathCoM(l,1,r),pathCoM(l,2,r),pathCoM(l,3,r),50)
                    %hold off;
  %              end
            end
            end
        end
    

File: Python/LBD/findPathCoM.py
import numpy as np
import pandas as pd

def findPathCoM(pathLUT, AtlasToPathLUT, AtlasCoM):
    """
    a function to find Center of Mass for Pathology

    Attributes
    ----------
    pathLUT : 
        
    AtlasToPathLUT : 
        
    AtlasCoM : 
       The Center of Mass for Atlas
    
    Returns
    ----------
    pathCoM : [NL x 3 x 2]
        Mean of Center of Mass (?)
    pathToAtlasIndex :

    """
    # Get number of anatomical regions to match to Atlas
    NL = len(AtlasToPathLUT)

    # numpy to save the Center of Mass of the pathology (anatomical region)
    pathCoM = np.zeros((NL, 3, 2))

    # List of matching anatomical region to atlas index
    pathToAtlasIndex = [[None, None] for _ in range(NL)]

    hemi = ['L', 'R']
    
    for l in range(NL):
        for r in range(2):
            # Get the matched Atlas region
            AOName = AtlasToPathLUT.Matched[l]

            # Get indexes of where pathname match Atlas region and {L, R} hemisphere
            # (modified csv file column name Label.Hemi --> Label_Hemi)
            pathLabelIdx = np.where((pathLUT.PathName.str.match(AOName) & pathLUT.Label_Hemi.str.match(hemi[r])))[0]

            if not (pathLabelIdx.size == 0): # If there is a match
                # Get Label_ID400x7 for the matched rows (modified csv file column name Label.ID400x7 --> Label_ID400x7)
                # Need to offset by 1 (index difference between matlab and python)
                atlasIdx = (pathLUT.Label_ID400x7.loc[pathLabelIdx] - 1).to_numpy()
                
                pathCoM[l,:,r] = np.nanmean(AtlasCoM[atlasIdx,:], axis=0)
                pathToAtlasIndex[l][r] = atlasIdx
    
    return pathCoM, pathToAtlasIndex
